File: client/common/network_client.py
"""
Network Client
Handles UDP communication with server
"""

import socket
from typing import Optional
from .byte_buffer import ByteBuffer
from .message_types import TIMEOUT_SECONDS, MAX_RETRIES, MAX_BUFFER_SIZE


class NetworkClient:
    """Handles network communication with the server."""
    
    def __init__(self, server_ip: str, server_port: int):
        self.server_ip = server_ip
        self.server_port = server_port
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.sock.settimeout(TIMEOUT_SECONDS)
        self.next_request_id = 1
    
    def get_next_request_id(self) -> int:
        """Get the next request ID and increment counter."""
        request_id = self.next_request_id
        self.next_request_id += 1
        return request_id
    
    def send_request(self, request_data: bytes, retries: int = MAX_RETRIES, timeout: Optional[float] = None) -> Optional[bytes]:
        """
        Send a request to the server and wait for a response.
        Implements retry logic for at-least-once semantics.
        
        Args:
            request_data: The request data to send
            retries: Number of retry attempts
            timeout: Optional custom timeout in seconds (uses default if None)
        """
        # Save original timeout
        original_timeout = self.sock.gettimeout()
        
        # Set custom timeout if provided
        if timeout is not None:
            self.sock.settimeout(timeout)
        
        try:
            for attempt in range(retries):
                try:
                    # Send request
                    self.sock.sendto(request_data, (self.server_ip, self.server_port))
                    
                    # Wait for response
                    response_data, _ = self.sock.recvfrom(MAX_BUFFER_SIZE)
                    return response_data
                    
                except socket.timeout:
                    if attempt < retries - 1:
                        print(f"Timeout, retrying... (attempt {attempt + 2}/{retries})")
                    else:
                        print("Request timeout after all retries")
                        return None
            
            return None
        finally:
            # Restore original timeout
            if timeout is not None:
                self.sock.settimeout(original_timeout)
    
    def close(self):
        """Close the socket."""
        self.sock.close()

File: java_client/src/main/java/com/facility/common/NetworkClient.java
package com.facility.common;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketTimeoutException;

/**
 * Network Client
 * Handles UDP communication with server
 */
public class NetworkClient {
    private String serverIp;
    private int serverPort;
    private DatagramSocket socket;
    private int nextRequestId;
    private double dropRate;

    public NetworkClient(String serverIp, int serverPort) {
        this(serverIp, serverPort, 0.0);
    }

    public NetworkClient(String serverIp, int serverPort, double dropRate) {
        this.serverIp = serverIp;
        this.serverPort = serverPort;
        this.dropRate = dropRate;
        this.nextRequestId = 1;
        
        try {
            this.socket = new DatagramSocket();
            this.socket.setSoTimeout(MessageTypes.TIMEOUT_SECONDS * 1000);
        } catch (Exception e) {
            throw new RuntimeException("Failed to create socket: " + e.getMessage());
        }
    }

    public String getServerIp() {
        return serverIp;
    }

    public int getServerPort() {
        return serverPort;
    }

    public synchronized int getNextRequestId() {
        return nextRequestId++;
    }

    public byte[] sendRequest(byte[] requestData) {
        return sendRequest(requestData, MessageTypes.MAX_RETRIES, null);
    }

    public byte[] sendRequest(byte[] requestData, int retries, Integer timeout) {
        // Save original timeout
        int originalTimeout = 0;
        try {
            originalTimeout = socket.getSoTimeout();
            
            // Set custom timeout if provided
            if (timeout != null) {
                socket.setSoTimeout(timeout * 1000);
            }

            for (int attempt = 0; attempt < retries; attempt++) {
                try {
                    // Simulate packet drop on client side
                    if (Math.random() < dropRate) {
                        System.out.println("[DROP] Request dropped (attempt " + 
                            (attempt + 1) + "/" + retries + ")");
                        // Simulate timeout
                        Thread.sleep(MessageTypes.TIMEOUT_SECONDS * 1000);
                        continue;
                    }

                    // Send request
                    InetAddress address = InetAddress.getByName(serverIp);
                    DatagramPacket sendPacket = new DatagramPacket(
                        requestData, requestData.length, address, serverPort);
                    socket.send(sendPacket);

                    // Wait for response
                    byte[] receiveBuffer = new byte[MessageTypes.MAX_BUFFER_SIZE];
                    DatagramPacket receivePacket = new DatagramPacket(
                        receiveBuffer, receiveBuffer.length);
                    socket.receive(receivePacket);

                    // Extract actual data
                    byte[] responseData = new byte[receivePacket.getLength()];
                    System.arraycopy(receiveBuffer, 0, responseData, 0, receivePacket.getLength());
                    return responseData;

                } catch (SocketTimeoutException e) {
                    if (attempt < retries - 1) {
                        System.out.println("Timeout, retrying... (attempt " + 
                            (attempt + 2) + "/" + retries + ")");
                    } else {
                        System.out.println("Request timeout after all retries");
                        return null;
                    }
                }
            }
            return null;
            
        } catch (Exception e) {
            System.err.println("Error sending request: " + e.getMessage());
            return null;
        } finally {
            // Restore original timeout
            if (timeout != null) {
                try {
                    socket.setSoTimeout(originalTimeout);
                } catch (Exception e) {
                    // Ignore
                }
            }
        }
    }

    public void close() {
        if (socket != null && !socket.isClosed()) {
            socket.close();
        }
    }
}
